gax/src/main/java/com/google/api/gax/tracing/TracedBatchedContextCallable.java

```diff
@@ -0,0 +1,118 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.tracing;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.InternalApi;
+import com.google.api.gax.batching.BatchedCallContext;
+import com.google.api.gax.rpc.ApiCallContext;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.api.gax.tracing.ApiTracerFactory.OperationType;
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+
+/**
+ * This callable wraps a batching callable chain in an {@link ApiTracer} and annotates {@link
+ * BatchedCallContext} batching context data.
+ *
+ * <p>For internal use only.
+ */
+@InternalApi("For internal use by google-cloud-java clients only")
+public class TracedBatchedContextCallable<RequestT, ResponseT>
+    extends UnaryCallable<RequestT, ResponseT> {
+
+  private final ApiTracerFactory tracerFactory;
+  private ApiCallContext baseCallContext;
+  private final SpanName spanName;
+  private final UnaryCallable<RequestT, ResponseT> innerCallable;
+
+  public TracedBatchedContextCallable(
+      UnaryCallable<RequestT, ResponseT> innerCallable,
+      ApiCallContext callContext,
+      ApiTracerFactory tracerFactory,
+      SpanName spanName) {
+    this.baseCallContext = Preconditions.checkNotNull(callContext);
+    this.tracerFactory = Preconditions.checkNotNull(tracerFactory);
+    this.spanName = Preconditions.checkNotNull(spanName);
+    this.innerCallable = Preconditions.checkNotNull(innerCallable);
+  }
+
+  /**
+   * Creates an {@link ApiTracer} and annotates batching context data. Performs a call
+   * asynchronously.
+   */
+  public ApiFuture<ResponseT> futureCall(RequestT request, BatchedCallContext batchedCallContext) {
+    ApiTracer tracer =
+        tracerFactory.newTracer(baseCallContext.getTracer(), spanName, OperationType.Batching);
+    TraceFinisher<ResponseT> finisher = new TraceFinisher<>(tracer);
+
+    try {
+      tracer.batchRequestThrottled(batchedCallContext.getTotalThrottledTimeMs());
+      tracer.batchRequestSent(
+          batchedCallContext.getElementCount(), batchedCallContext.getByteCount());
+      baseCallContext = baseCallContext.withTracer(tracer);
+      ApiFuture<ResponseT> future = innerCallable.futureCall(request, baseCallContext);
+      ApiFutures.addCallback(future, finisher, MoreExecutors.directExecutor());
+
+      return future;
+    } catch (RuntimeException e) {
+      finisher.onFailure(e);
+      throw e;
+    }
+  }
+
+  /** Calls the wrapped {@link UnaryCallable} within the context of a new trace. */
+  @Override
+  public ApiFuture futureCall(RequestT request, ApiCallContext context) {
+    ApiCallContext mergedContext = baseCallContext.merge(context);
```

Comment:
Looks like there is a lot of duplication between these two futureCall methods. Can you please rewrite them such that the code is reused, instead of being duplicated?

Reply:
Refactored the methods.